File: ts/gateway/deployment_audit.ts
#!/usr/bin/env node
/**
 * Independently audit Cloud API Gateway and model-serving deployments.
 *
 * This command deliberately has no completion signature or response bytes.
 * It establishes that the returned attestations are fresh and valid deployment
 * evidence, but it does not associate either deployment with a chat response.
 */

import * as crypto from 'crypto';

import {
  type SigningAlgo,
  fetchGatewayAttestation,
  fetchModelAttestations,
} from './cloud_api';
import {
  type VerifiedGatewayAttestation,
  verifyGatewayAttestation,
  verifyModelAttestation,
} from './attestation';
import { normalizeSigningAlgo } from '../common/dstack_attestation';

export interface AuditModelAttestationsParams {
  model: string;
  nonce?: string;
  signingAlgo?: SigningAlgo;
}

export interface AuditGatewayAttestationParams {
  nonce?: string;
  signingAlgo?: SigningAlgo;
}

/**
 * Fetch and independently audit Gateway evidence. This verifies the Gateway
 * deployment and its evidence-request TLS peer, not a particular completion.
 */
export async function auditGatewayAttestation({
  nonce = crypto.randomBytes(32).toString('hex'),
  signingAlgo,
}: AuditGatewayAttestationParams = {}): Promise<VerifiedGatewayAttestation> {
  const evidence = await fetchGatewayAttestation({ nonce, signingAlgo });
  return verifyGatewayAttestation({
    attestation: evidence.attestation,
    nonce,
    peerSpkiFingerprint: evidence.peerSpkiFingerprint,
  });
}

/**
 * Fetch and independently audit every NEAR model attestation for a model.
 * The result is deployment evidence only; no completion is involved.
 */
export async function auditModelAttestations({
  model,
  nonce = crypto.randomBytes(32).toString('hex'),
  signingAlgo,
}: AuditModelAttestationsParams): Promise<void> {
  const attestations = await fetchModelAttestations({ model, nonce, signingAlgo });
  for (const [index, attestation] of attestations.entries()) {
    console.log(
      `\n========================================\nModel attestation ${index + 1}\n========================================`,
    );
    await verifyModelAttestation({ attestation, nonce });
  }
}

async function main(): Promise<void> {
  const args = process.argv.slice(2);
  const modelIndex = args.indexOf('--model');
  const model =
    modelIndex !== -1 && args[modelIndex + 1]
      ? args[modelIndex + 1]
      : 'deepseek-ai/DeepSeek-V3.1';
  const algoIndex = args.indexOf('--signing-algo');
  const signingAlgo =
    algoIndex !== -1 && args[algoIndex + 1]
      ? normalizeSigningAlgo(args[algoIndex + 1]!)
      : undefined;

  if (!process.env.API_KEY) {
    throw new Error('API_KEY is required to fetch Cloud API attestations');
  }

  console.log('========================================');
  console.log('NEAR AI Cloud Gateway deployment audit');
  console.log('========================================');

  await auditGatewayAttestation({ signingAlgo });
  await auditModelAttestations({ model, signingAlgo });
}

if (require.main === module) {
  main().catch((error) => {
    console.error('\nVerification failed:', error instanceof Error ? error.message : error);
    process.exitCode = 1;
  });
}

File: py/gateway/deployment_audit.py
#!/usr/bin/env python3
"""Independently audit Cloud API Gateway and model-serving deployments.

This command accepts no completion signature or response bytes. It verifies
fresh deployment evidence but cannot associate either attestation with a chat.
"""

from __future__ import annotations

import argparse
import asyncio
from typing import Optional, Sequence

from py.common.dstack_attestation import SUPPORTED_SIGNING_ALGOS
from py.gateway.cloud_api import (
    API_KEY,
    fetch_gateway_attestation,
    fetch_model_attestations,
)
from py.gateway.attestation import (
    VerifiedGatewayAttestation,
    VerifiedModelAttestation,
    verify_gateway_attestation,
    verify_model_attestation,
)


async def audit_gateway_attestation(
    signing_algo: Optional[str] = None,
) -> VerifiedGatewayAttestation:
    """Fetch and independently audit Gateway evidence, without a completion."""

    attestation, nonce, peer_spki = fetch_gateway_attestation(signing_algo=signing_algo)
    return await verify_gateway_attestation(attestation, nonce, peer_spki)


async def audit_model_attestations(
    model: str, signing_algo: Optional[str] = None
) -> Sequence[VerifiedModelAttestation]:
    """Fetch and independently audit model evidence, without a completion."""

    attestations, nonce = fetch_model_attestations(model, signing_algo=signing_algo)
    verified: list[VerifiedModelAttestation] = []
    for index, attestation in enumerate(attestations, start=1):
        print(
            f"\n========================================\n"
            f"Model attestation {index}\n"
            f"========================================"
        )
        verified.append(await verify_model_attestation(attestation, nonce))
    return verified


async def main() -> int:
    parser = argparse.ArgumentParser(
        description="Audit NEAR AI Cloud Gateway and model attestations"
    )
    parser.add_argument("--model", default="deepseek-ai/DeepSeek-V3.1")
    parser.add_argument(
        "--signing-algo",
        choices=sorted(SUPPORTED_SIGNING_ALGOS),
        help="Request a specific algorithm; omit to use the Cloud API default.",
    )
    args = parser.parse_args()

    if not API_KEY:
        raise RuntimeError("API_KEY is required to fetch Cloud API attestations")

    print("========================================")
    print("NEAR AI Cloud Gateway deployment audit")
    print("========================================")
    await audit_gateway_attestation(args.signing_algo)
    await audit_model_attestations(args.model, args.signing_algo)
    return 0


if __name__ == "__main__":
    raise SystemExit(asyncio.run(main()))
